structuredheadings/plugin.js

```diff
@@ -1,67 +1,139 @@
 (function () {
   CKEDITOR.plugins.add("structuredheadings", {
-    icons: "structuredheadings",
+    icons: "autonumberheading, matchheading",
     init: function (editor) {
       // list of elements allowed to be numbered
       var allowedElements = ["h1", "h2", "h3", "h4", "h5", "h6"];
 
       editor.addContentsCss(this.path + "styles/numbering.css");
 
       // Helper function to set button state
-      var setCommandState = function (state) {
+      var setCommandState = function (command, state) {
         switch (state) {
         case "on":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_ON);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_ON);
           break;
         case "off":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_OFF);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_OFF);
           break;
         case "disabled":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_DISABLED);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_DISABLED);
           break;
         }
 
       };
 
+      var isNumbered = function (element) {
+        if (element.hasClass("autonumber")) {
+          return true;
+        } else {
+          return false;
+        }
+      };
+
       // The command to add an autonumbering class to selection
       editor.addCommand("autoNumberHeading", {
         allowedContent: "h1(*); h2(*); h3(*); h4(*); h5(*); h6(*)",
-        requiredContent: allowedElements.join(";"),
         startDisabled: true,
         exec: function () {
           var element = editor.getSelection().getStartElement();
 
-          if (!element.hasClass("autonumber")) {
+          if (!isNumbered(element)) {
             element.addClass("autonumber");
-            setCommandState("on");
+            setCommandState("autoNumberHeading", "on");
           } else {
             element.removeClass("autonumber");
-            setCommandState("off");
+            setCommandState("autoNumberHeading", "off");
           }
         }
       });
 
+      editor.addCommand("matchHeading", {
+        startDisabled: true,
+        exec: function () {
+          var element = editor.getSelection().getStartElement();
```

Comment:
Ok so, actually to my surprise case 1 (two headings with a selection spanning both) the ToggleHeading functionality applies to both headings, including setting autonumber on both. However the AutoNumber functionality will only work on the first heading at the start of the selection. The fact that ToggleHeading works on both is due to the editor.applyStyle covering the full selection. Will have to give that some more though on what is appropriate.

Case 2, the bold text as expected does not work. Both commands are disabled inside of a strong tag. I could probably check that the parent is a header, but then in the case of bold-italic, I'd have to check grandparent, and how would I know how far back to look without specifically checking if I am in a certain tag. I could getPrevious recursively to get the last instance of a Header but I'm not sure that would always lead to an expected result, would have to try it out.